lib/commons/dom/is-visible.js

```diff
@@ -13,14 +13,187 @@ const clipRegex = /rect\s*\(([0-9]+)px,?\s*([0-9]+)px,?\s*([0-9]+)px,?\s*([0-9]+
 const clipPathRegex = /(\w+)\((\d+)/;
 
 /**
- * Determines if an element is hidden with a clip or clip-path technique
- * @method isClipped
+ * Determine whether an element is visible
+ * @method isVisible
  * @memberof axe.commons.dom
- * @private
- * @param  {CSSStyleDeclaration} style Computed style
- * @return {Boolean}
+ * @instance
+ * @param {HTMLElement} el The HTMLElement
+ * @param {Boolean} screenReader When provided, will evaluate visibility from the perspective of a screen reader
+ * @param {Boolean} recursed
+ * @return {Boolean} The element's visibilty status
+ */
+export default function isVisible(el, screenReader, recursed) {
+  if (!el) {
+    throw new TypeError(
+      'Cannot determine if element is visible for non-DOM nodes'
+    );
+  }
+
+  const vNode = el instanceof AbstractVirtualNode ? el : getNodeFromTree(el);
+  el = vNode ? vNode.actualNode : el;
+
+  const cacheName = '_isVisible' + (screenReader ? 'ScreenReader' : '');
+  const { DOCUMENT_NODE, DOCUMENT_FRAGMENT_NODE } = window.Node ?? {};
+  const nodeType = vNode ? vNode.props.nodeType : el.nodeType;
+  const nodeName = vNode ? vNode.props.nodeName : el.nodeName.toLowerCase();
+
+  if (vNode && typeof vNode[cacheName] !== 'undefined') {
+    return vNode[cacheName];
+  }
+
+  if (nodeType === DOCUMENT_NODE) {
+    return true;
+  }
+
+  const hiddenMethods = [elementHidden];
+  let style;
+  if (screenReader) {
+    hiddenMethods.push(ariaHidden);
+  }
+
+  if (el) {
+    if (nodeType === DOCUMENT_FRAGMENT_NODE) {
+      el = el.host; // grab the host Node
+    }
+
+    style = window.getComputedStyle(el, null);
+    if (style === null) {
+      return false;
+    }
+
+    /**
+     * check visibility of `AREA`
+     * Note:
+     * Firefox's user-agent always sets `AREA` element to `display:none`
+     * hence excluding the edge case, for visibility computation
+     */
+    if (nodeName === 'area') {
+      return isAreaVisible(el, screenReader, recursed);
+    }
+
+    hiddenMethods.push(displayHidden);
+
+    if (!screenReader) {
+      hiddenMethods.push(
+        opacityHidden,
+        clipHidden,
+        scrollHidden,
+        overflowHidden
+      );
+    }
+    if (!recursed) {
+      hiddenMethods.push(visibilityHidden);
+    }
+    if (!screenReader && !recursed) {
+      hiddenMethods.push(offscreenHidden);
+    }
+  }
+
+  if (
+    hiddenMethods.some(method =>
+      method({
+        el,
+        vNode,
+        nodeName,
+        style
+      })
+    )
+  ) {
+    return false;
+  }
+
+  const parent = el
+    ? el.assignedSlot
+      ? el.assignedSlot
+      : el.parentNode
+    : vNode.parent;
+  let visible = el ? false : true;
+  if (parent) {
+    visible = isVisible(parent, screenReader, true);
+  }
```

Comment:
This looks kinda odd. How about: 
```suggestion
  let visible;
  if (parent) {
    visible = isVisible(parent, screenReader, true);
  } else {
    visible = el ? false : true
  }
```